SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -351,19 +358,20 @@ public static void Init()
         }
 
 
-        static MachineID GenerateMachineID()
+        static MachineID GenerateMachineID(object state)
         {
             // the aug 25th 2015 CM update made well-formed machine MessageObjects required for logon
             // this was flipped off shortly after the update rolled out, likely due to linux steamclients running on distros without a way to build a machineid
             // so while a valid MO isn't currently (as of aug 25th) required, they could be in the future and we'll abide by The Valve Law now
 
-            var machineId = new MachineID();
+            var provider = (IMachineInfoProvider)state;
 
-            MachineInfoProvider provider = MachineInfoProvider.GetProvider();
+            var machineId = new MachineID();
 
-            machineId.SetBB3( GetHexString( provider.GetMachineGuid() ) );
-            machineId.SetFF2( GetHexString( provider.GetMacAddress() ) );
-            machineId.Set3B3( GetHexString( provider.GetDiskId() ) );
+            // Custom implementations can fail for any particular field, in which case we fall back to DefaultMachineInfoProvider.
+            machineId.SetBB3( GetHexString( provider.GetMachineGuid() ?? DefaultMachineInfoProvider.Instance.GetMachineGuid() ) );
```

Comment:
Do we want to handle providers throwing exceptions?

With this current code it's looking like this will throw when the task is `Wait`ed, which will manifest itself during `SteamUser/SteamGameServer.Logon`.

Do we want to bubble up that exception there or fall back to the default provider?

Reply:
I think for SteamKit internals it is best to play it safe and quietly handle failure, but for consumer-supplied implementations it should be obvious and in their face that their component exploded.